From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tomas Bzatek <tbzatek@redhat.com>
Date: Mon, 22 Mar 2021 19:22:56 +0100
Subject: [PATCH] part: Fix elements leak in bd_part_get_part_by_pos()

Breaking the cycle leaves the rest of the elements in the air.
---
 src/plugins/part.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/plugins/part.c b/src/plugins/part.c
index b7e92bd..6af181e 100644
--- a/src/plugins/part.c
+++ b/src/plugins/part.c
@@ -872,16 +872,18 @@ BDPartSpec* bd_part_get_part_by_pos (const gchar *disk, guint64 position, GError
             if ((*parts_p)->type == BD_PART_TYPE_EXTENDED) {
                 /* we don't want to return extended partition here -- there
                    is either another logical or free space at this position */
-                bd_part_spec_free (*parts_p);
                 continue;
             }
 
             ret = *parts_p;
             break;
-        } else
-            bd_part_spec_free (*parts_p);
+        }
     }
 
+    /* free the array except the selected element */
+    for (BDPartSpec **parts_p = parts; *parts_p; parts_p++)
+        if (*parts_p != ret)
+            bd_part_spec_free (*parts_p);
     g_free (parts);
 
     return ret;
-- 
2.33.1

